### ISSUE TITLE
[clang] fails to compile valid code involving `variant` and `unordered_map` with libstdc++ version <12

### ISSUE THREAD

**Issue by toughengineer** [llvm/llvm-project#84487]
# tl;dr:
clang, starting from version 7.0.1 all the way up to the current trunk, in C++17 mode fails to compile the following code when combined with libstdc++ version <12:
https://godbolt.org/z/vrKc9877v
<details><summary>(click/tap to expand, this is the stripped minimal example that I could come up with that reproduces the problem)</summary>

```cpp
#include <vector>
#include <unordered_map>
#include <variant>
#include <type_traits>
#include <cstdint> // for uint8_t

namespace detail {
  template<typename Allocator, typename T>
  using ReboundAllocator = typename std::allocator_traits<Allocator>::template rebind_alloc<T>;
  template<typename T>
  using RemoveCVRef = std::remove_cv_t<std::remove_reference_t<T>>;
}

template<typename Allocator>
struct BasicValue {
  using Null = std::monostate;
  using Object = std::unordered_map<int, BasicValue,
    std::hash<int>, std::equal_to<int>,
    detail::ReboundAllocator<Allocator, std::pair<const int, BasicValue>>>;
  using Variant = std::variant<Null, Object>;

  BasicValue() = default;
  template<typename T,
    std::enable_if_t<!std::is_same_v<detail::RemoveCVRef<T>, BasicValue> &&
    std::is_convertible_v<T&&, Variant>, int> = 0>
  BasicValue(T &&v) noexcept(std::is_nothrow_constructible_v<Variant, T&&>) : m_data{ std::forward<T>(v) } {}

private:
  struct Workaround final {
    Workaround() noexcept { ::new(&storage) Variant{}; }
    template<typename... T>
    Workaround(T&&... v) noexcept(std::is_nothrow_constructible_v<Variant, T&&...>) {
      ::new(&storage) Variant{ std::forward<T>(v)... };
    }
    Workaround(const Workaround &other) { ::new(&storage) Variant{ other.operator const Variant & () }; }
    ~Workaround() {
      static_assert(sizeof(DummyVariant) == sizeof(Variant));
      static_assert(alignof(DummyVariant) == alignof(Variant));
      operator Variant&().~Variant();
    }
    auto &operator=(const Workaround &other) { return operator Variant & () = other.operator const Variant & (); }
    operator const Variant&() const& noexcept { return *std::launder(reinterpret_cast<const Variant*>(&storage)); }
    operator Variant&() & noexcept { return *std::launder(reinterpret_cast<Variant*>(&storage)); }

  private:
    using DummyUnorderedMap = std::unordered_map<int, int,
      std::hash<int>, std::equal_to<int>,
      detail::ReboundAllocator<Allocator, std::pair<const int, int>>>;
    using DummyVariant = std::variant<Null, DummyUnorderedMap>;
    alignas(DummyVariant) uint8_t storage[sizeof(DummyVariant)]{};
  } m_data;
};

// for some reason clang needs this
//static_assert(std::is_copy_constructible_v<BasicValue<std::allocator<char>>>);

using Value = BasicValue<std::allocator<char>>;


struct ShouldBeIrrelevant {
  // commenting any of the following lines gives:
  // "error: incomplete type 'std::unordered_map<...>"
  Value value;
  std::vector<int> bar;
};


int main() {
  Value v;
  const auto c = std::unordered_map<int, Value>{}; 

  //error: call to implicitly-deleted copy constructor of 'std::pair<const int, BasicValue<std::allocator<char>>>'
  v = c;
}
```
</details>
giving error

    error: call to implicitly-deleted copy constructor of 'std::pair<const int, BasicValue<std::allocator<char>>>'

while other compilers have no problems with this code, namely GCC and MSVC.

# Problem
## C++17 mode
Something breaks in clang (practically all versions supporting C++17 "enough", i.e. >7.0.1) with this code when it is used with libstdc++ version <12 (in particular versions 9.2..11.4, earlier versions do not support C++17 "enough")

If you remove the part
```cpp
struct ShouldBeIrrelevant {
  Value value;
  std::vector<int> bar;
};
```
or one or both of the members clang fails with error:

    error: incomplete type 'std::unordered_map<int, BasicValue<std::allocator<char>>>' used in type trait expression

though it seems completely irrelevant to the code that causes the error:
```cpp
Value v;
const auto c = std::unordered_map<int, Value>{};
v = c;
```

## Later standard modes
With e.g. C++20 or C++23, the error is spelled differently but is practically the same:

    error: no matching function for call to 'construct_at'

# Workaround
If I add a `static_assert`
```cpp
static_assert(std::is_copy_constructible_v<BasicValue<std::allocator<char>>>);
```
everythng starts to work as expected, and the `struct ShouldBeIrrelevant` snippet becomes truly irrelevant, i.e. removing it completely or removing any or both of the members has no effect.

# Context
In libstdc++ versions <12 `std::unordered_map` does not support incomplete types, we can work around that by using byte array storage, placement `::new()` to construct an object in that storage, and `reinterpret_cast` to turn it to a pointer to desired type inside implementation where the type is complete.

We _guess_ that alignment and size of some complete `std::unordered_map` instantiation is the same as the one with currently incomplete type, and `static_assert` that we guessed correctly in the destructor.

**Comment by cor3ntin** [comment]
@toughengineer, do you have a self-contained reproduction? (the preprocessed output would be enough)

**Comment by toughengineer** [comment]
@cor3ntin, it seems that the problem does not reproduce with the current trunk that is currently deployed to the Compiler Explorer
https://godbolt.org/z/vrKc9877v (link from the initial message, ensure that the compiler is set to "x86-64 clang (trunk)")

It reproduces with the latest release version 20.1 that is deployed to the Compiler Explorer though.

You can find self contained reproduction code in the initial message, as well as here:
https://godbolt.org/z/vrKc9877v (the same link from the initial message)

Please let me know if you need me to produce the preprocessed output for that code for you (I'll need a few days to get to it).

**Comment by cor3ntin** [comment]
@toughengineer Please do. We can't know if the bug is fixed because clang changed, or because the libstdc++ changed. 
But if you are satisfied that the bug is fixed on clang trunk we can close.

Thanks

**Comment by toughengineer** [comment]
@cor3ntin, if you follow
https://godbolt.org/z/vrKc9877v (link from the initial message)
version of libstdc++ is indeed 11.4 (`__GLIBCXX__ == 20230528`) which indicates that libstdc++ version did not change, and only Clang trunk changed.

I will be satisfied if/when released version of Clang successfully compiles the code from this issue with libstc++ vesrion <12, i.e. the initial issue is fixed (whatever it is).

I will produce preprocessed output (if it is still needed) until the end of the next week.

**Comment by toughengineer** [comment]
@cor3ntin, here's the preprocessed output of Clang 20.1 as deployed at Compiler Explorer:
[preprocessed.zip](https://github.com/user-attachments/files/20177388/preprocessed.zip)

And here's the preprocessed output for C++20 mode:
[preprocessed_c++20.zip](https://github.com/user-attachments/files/20177513/preprocessed_c%2B%2B20.zip)

See the link from the initial message:
https://godbolt.org/z/vrKc9877v

As well, you can find there the compiler flags that the error reproduces with:
```
-gdwarf-4 -g -o output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-11.4.0 -fcolor-diagnostics -fno-crash-diagnostics -std=c++17 example.cpp
```

And for C++20 mode:
```
-gdwarf-4 -g -o output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-11.4.0 -fcolor-diagnostics -fno-crash-diagnostics -std=c++20 example.cpp
```

The relevant parts are probably just
```
--gcc-toolchain=/opt/compiler-explorer/gcc-11.4.0 -std=c++17
```
and
```
--gcc-toolchain=/opt/compiler-explorer/gcc-11.4.0 -std=c++20
```
respectively.
